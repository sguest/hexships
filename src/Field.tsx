import * as React from 'react'
import { useEffect, useRef } from 'react';
import * as hexUtils from './hex-utils';
import GameSettings from './config/GameSettings';

export interface FieldProps {
    settings: GameSettings
}

export default function Field(props: FieldProps) {
    const canvasRef = useRef<HTMLCanvasElement>(null);
    const { settings } = props;
    useEffect(() => {
        const canvas = canvasRef.current;
        if(!canvas) {
            return;
        }
        const context = canvas.getContext('2d');
        if(!context) {
            return;
        }
        context.clearRect(0, 0, canvas.width, canvas.height);
        context.strokeStyle = 'blue';
        context.beginPath();
        const cells = hexUtils.getGridCells(settings.gridSize);
        for(let cell of cells) {
            const corners = hexUtils.getCorners(cell.x, cell.y, 20);
            context.moveTo(corners[5].x + 200, corners[5].y + 250);
            for(let corner of corners) {
                context.lineTo(corner.x + 200, corner.y + 250);
            }
        }
        context.stroke();

    }, [settings, canvasRef])
    return <canvas ref={canvasRef} width="500" height="500" />
}