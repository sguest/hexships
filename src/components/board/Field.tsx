import { useEffect, useRef } from 'react';
import * as hexUtils from '../../utils/hex-utils';
import { Point } from '../../utils/point-utils';
import * as pointUtils from '../../utils/point-utils';
import useScaledCanvas from './useScaledCanvas';
import { createUseStyles } from 'react-jss';

export interface FieldProps {
    gridSize: number
    uiScale: number
    gridDimensions: Point
}

const useStyles = createUseStyles({
    '@keyframes boardGradient': {
        from: { backgroundPosition: '0% 50%' },
        to: { backgroundPosition: '100% 50%' },
    },
    canvas: {
        animationName: '$boardGradient',
        animationDuration: '20s',
        animationIterationCount: 'infinite',
        animationTimingFunction: 'ease',
        animationDirection: 'alternate',
        position: 'absolute',
        zIndex: 1,
        pointerEvents: 'none',
        background: 'linear-gradient(80deg, #00545c 0%, #00405c 30%, #00545c 60%, #00405c 90%)',
        backgroundSize: '400% 400%',
        border: '5px solid #242f40',
        boxSizing: 'border-box',
    },
})

export default function Field(props: FieldProps) {
    const canvasRef = useRef<HTMLCanvasElement>(null);
    const classes = useStyles();

    useEffect(() => {
        useScaledCanvas(canvasRef, props.uiScale, context => {
            context.strokeStyle = 'black';
            context.beginPath();
            const cells = hexUtils.getGridCells(props.gridSize);
            for(const cell of cells) {
                const corners = hexUtils.getCorners(cell, props.gridSize);
                pointUtils.moveTo(context, pointUtils.add(corners[5], pointUtils.multiplyScalar(props.gridDimensions, 0.5)));
                for(const corner of corners) {
                    pointUtils.lineTo(context, pointUtils.add(corner, pointUtils.multiplyScalar(props.gridDimensions, 0.5)));
                }
            }
            context.stroke();
        });
    }, [props.gridSize, canvasRef, props.uiScale, props.gridDimensions])

    return <canvas ref={canvasRef} width={props.gridDimensions.x * props.uiScale} height={props.gridDimensions.y * props.uiScale} className={classes.canvas} data-testid="field-canvas" />
}