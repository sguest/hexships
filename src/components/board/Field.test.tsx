import 'jest-canvas-mock';
import { render, screen } from '@testing-library/react'
import Field from './Field'

test('Should render grid lines', () => {
    render(<Field uiScale={1} gridSize={1} gridDimensions={{ x: 1, y: 1 }} />)
    const canvas = screen.getByTestId('field-canvas') as HTMLCanvasElement;
    const context = canvas.getContext('2d');
    const events = context!.__getEvents();
    expect(events).toMatchSnapshot();
})