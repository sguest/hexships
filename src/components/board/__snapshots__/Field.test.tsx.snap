// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render grid lines 1`] = `
Array [
  Object {
    "props": Object {
      "a": 1,
      "b": 0,
      "c": 0,
      "d": 1,
      "e": 0,
      "f": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "setTransform",
  },
  Object {
    "props": Object {
      "height": 1,
      "width": 1,
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  Object {
    "props": Object {
      "x": 1,
      "y": 1,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "scale",
  },
  Object {
    "props": Object {
      "value": "#000000",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "strokeStyle",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "beginPath",
  },
  Object {
    "props": Object {
      "x": 10.499999999999986,
      "y": -16.82050807568878,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "moveTo",
  },
  Object {
    "props": Object {
      "x": 20.5,
      "y": 0.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineTo",
  },
  Object {
    "props": Object {
      "x": 10.500000000000002,
      "y": 17.82050807568877,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineTo",
  },
  Object {
    "props": Object {
      "x": -9.499999999999996,
      "y": 17.820508075688775,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineTo",
  },
  Object {
    "props": Object {
      "x": -19.5,
      "y": 0.5000000000000024,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineTo",
  },
  Object {
    "props": Object {
      "x": -9.500000000000009,
      "y": -16.82050807568877,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineTo",
  },
  Object {
    "props": Object {
      "x": 10.499999999999986,
      "y": -16.82050807568878,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineTo",
  },
  Object {
    "props": Object {
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "x": 10.499999999999986,
            "y": -16.82050807568878,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "moveTo",
        },
        Object {
          "props": Object {
            "x": 20.5,
            "y": 0.5,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "lineTo",
        },
        Object {
          "props": Object {
            "x": 10.500000000000002,
            "y": 17.82050807568877,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "lineTo",
        },
        Object {
          "props": Object {
            "x": -9.499999999999996,
            "y": 17.820508075688775,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "lineTo",
        },
        Object {
          "props": Object {
            "x": -19.5,
            "y": 0.5000000000000024,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "lineTo",
        },
        Object {
          "props": Object {
            "x": -9.500000000000009,
            "y": -16.82050807568877,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "lineTo",
        },
        Object {
          "props": Object {
            "x": 10.499999999999986,
            "y": -16.82050807568878,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "lineTo",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "stroke",
  },
]
`;
